src/Highlighter.ts

```diff
@@ -144,13 +144,40 @@ export default class Highlighter {
     return this.container.ownerDocument;
   }
 
+  private isImgOrIframe = (node: Node) => node.nodeName === 'IMG' || node.nodeName === 'IFRAME';
+
   private snapSelection = () => {
     const selection = this.document.getSelection();
 
     if (!selection || selection.isCollapsed) {
       return;
     }
 
+    const anchor = selection.anchorNode;
+    const focus = selection.focusNode;
+
+    if (anchor && focus) {
+      const anchorParent = anchor.parentNode;
+      const focusParent = focus.parentNode;
+
+      // if selection begins on but doesn't end on img/iframe
+      if (this.isImgOrIframe(anchor) && anchorParent && !this.isImgOrIframe(focus)) {
+        selection.setBaseAndExtent(anchorParent, selection.anchorOffset, focus, selection.focusOffset);
+      }
+
+      // if selection ends on img/iframe
+      if (this.isImgOrIframe(focus) && focusParent) {
+        const focusGrandparent = focusParent.parentNode;
+        const focusIsImgWithCaption = focus.nodeName === 'IMG' && focusGrandparent && focusGrandparent.nextSibling && (focusGrandparent.nextSibling as Element).className === 'os-caption-container';
+
+        // for images with captions, set caption as new focusNode, else use focus parent
+        const newFocus = focusGrandparent && focusGrandparent.nextSibling && focusIsImgWithCaption ? focusGrandparent.nextSibling : focusParent;
+        const newFocusOffset = focusIsImgWithCaption ? selection.focusOffset : selection.focusOffset + 1;
+        const newAnchor = this.isImgOrIframe(anchor) && anchorParent ? anchorParent : anchor;
+        selection.setBaseAndExtent(newAnchor, selection.anchorOffset, newFocus, newFocusOffset);
+      }
```

Comment:
@TomWoodward @jarosik10 This is not pretty. Part of the goal of lines 169–177 is to allow a user to end a selection on an image in Firefox—currently if you try to do this in production, the selection snaps to the first word after the image. I had indicated that there was a more efficient solution, but that was wrong—the only thing I've found that works is to mimic Chrome and use a nearby sibling as the new `focusNode`, in this case the `os-caption-container`. 

@TomWoodward pointed out that not every image has a caption (see practice problems [here](https://openstax.org/books/biology-ap-courses/pages/1-test-prep-for-ap-r-courses) and [here](https://openstax.org/books/physics/pages/2-3-position-vs-time-graphs)), so there will be a handful of cases that are not covered.

A side effect of setting the caption as the new focusNode is that the EditCard drops down beneath the caption (not an issue on Chrome because Chrome does not include `range.endContainer.height` when calculating `getBoundingClientRect().bottom`). To solve this we could check for Firefox and then toggle a property on the highlight that indicates that `range.endContainer.height` should be subtracted during offset calculations. This works but I don't know if it's worth it.

Note: the reported [bug](https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/1619) only mentions starting a highlight on an image, so I've gone down the rabbit hole a bit here. That bug is resolved on lines 164–166.

Reply:
I still have problems with images at the end of the selection. It's really tricky. However, the bug solving part 164-166 works great.